force-app/main/default/classes/DMLOperations.cls

```diff
@@ -145,66 +154,142 @@ public with sharing class DMLOperations {
      * Example Contact.LastName = 'Doe' Link to Account.Name = 'Doe' 
      * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
-	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
-	 */
-	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
-	}
-
+     * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
+     */
+    public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+        // Loop through each Contact
+        List<String> lastNames = new List<String>(); //create a list to store the last names
+        for(Contact con : contactsToUpsert){
+            lastNames.add(con.LastName); //loop through the contacts provided and put the last names in the list
+        }
+
+        List<Account> existingAccounts = [SELECT Id, Name FROM Account WHERE Name IN :lastNames]; //list of existing Accounts that match our last names
+        
+        Map<String, Id> accountNameToId = new Map<String, Id>(); //create a map of account names and Ids
+        for(Account a : existingAccounts) {
+            accountNameToId.put(a.Name, a.Id); 
+        }
+
+        List<Account> newAccounts = new List<Account>(); //create new accounts for missing last names
+        for(String lastName : lastNames) {
+            if(!accountNameToId.containsKey(lastName)) {
+            newAccounts.add(new Account(Name = lastName));
+            }
+        }
+
+        insert newAccounts;
+        for(Account a : newAccounts) {
+            accountNameToId.put(a.Name, a.Id);
+        }
+    
+        for(Contact c : contactsToUpsert) { // Link the Contact to its Account using the Account Id
+            c.AccountId = accountNameToId.get(c.LastName);
+        }
+    
+        upsert contactsToUpsert; // Upsert the Contacts
+    }
     /**
      * Question 11
      * Insert a list of Leads and then deletes them.
      * @param leadNames The list of Lead names to be inserted.
      */
     public static void insertAndDeleteLeads(List<String> leadNames) {
+        List<Lead> myLeads = new List<Lead>();
         // Loop through each Lead name
-
+        for(String leadName : leadNames){
+            Lead newLead = new Lead(
+                LastName = leadName, 
+                Company = 'Oy and Chroma and Milli Beagles Inc' //Company is a required field, now my dogs exist in my code!
+            ); 
+            myLeads.add(newLead);
+        }
         // Insert the Leads
-
+        if (!myLeads.isEmpty()) { //this checks if the list of leads is not empty, if it returns false, insert them
+            insert myLeads;
```

Comment:
While it's not a bad idea to check for empty, null, etc., Apex actually ignores any DML statement on an empty list. So, in these cases, you can skip it:
```suggestion
        insert myLeads;
```